src/main/scala/strawman/collection/immutable/StrictOptimizedSeqOps.scala

```diff
@@ -11,6 +11,19 @@ trait StrictOptimizedSeqOps[+A, +CC[_], +C]
   extends SeqOps[A, CC, C]
     with StrictOptimizedIterableOps[A, CC, C] {
 
+  override def distinct: C = {
+    val builder = newSpecificBuilder()
+    val seen = mutable.HashSet.empty[A]
+
+    for (x <- coll) {
+      if (!seen.contains(x)) {
+        seen += x
+        builder += x
+      }
+    }
+    builder.result()
+  }
```

Comment:
You can create a `strawman.collection.StrictOptimizedSeqOps` trait (non `immutable`) if some mutable collections can benefit from this optimization (I guess, `ArrayBuffer` and `ListBuffer` only).